subsys/net/lib/sockets/sockets.c

```diff
@@ -706,11 +707,13 @@ static int send_check_and_wait(struct net_context *ctx, int status,
 					  K_POLL_MODE_NOTIFY_ONLY,
 					  net_tcp_tx_sem_get(ctx));
 
-			k_poll(&event, 1, WAIT_BUFS);
+			k_poll(&event, 1, *retry_timeout);
 		} else {
-			k_sleep(WAIT_BUFS);
+			k_sleep(*retry_timeout);
 		}
 	}
+	/* Exponentially increase the retry timeout */
+	retry_timeout->ticks <<= 1;
```

Comment:
I don't think external code should mess with `k_timeout_t` internals, even the structure, even the structure docs somewhat prohibit that (`Applications should not inspect the internal data once constructed.`). 

If there's no suitable API to double the timeout value, perhaps we should operate on an integer and convert it to `k_timeout_t` on the fly.